### PR TITLE
Add '--listen-address' option and fix default environment for running…

### PR DESCRIPTION
**Pull request by colinmollenhour**
… as Docker container. Separates listen address from advertised hostname.

When running in docker and exposing the port if you use the defaults it will listen on the containers localhost which cannot be reached from the host, it needs to listen on 0.0.0.0. However if you set the hostname to 0.0.0.0 you get a url reported in the console that is invalid. This lets you separate the listen address from the reported url hostname.

If anything since this is a dev tool I'd just remove the hostname option entirely and always report the hostname as "localhost", but without a patch it is kinda annoying that the url reported in the console cannot be visited and then you have to determine/remember/figure out the correct url to visit.

This patch causes the most commonly needed defaults to be used whether it is run from the host directly using node/npm or if it is run from a docker container. Needing to change the defaults would be pretty rare.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -4,7 +4,8 @@ LABEL maintainer="Daniel Schröder <daniel.schroeder@skriptfabrik.com>"
 
 ARG ELEMENTS_CLI_VERSION=latest
 
-ENV ELEMENTS_HOSTNAME=0.0.0.0
+ENV ELEMENTS_HOSTNAME=localhost
+ENV ELEMENTS_LISTEN_ADDRESS=0.0.0.0
```
Comment (schroedan) [inline Dockerfile:8]:
```suggestion
ENV ELEMENTS_HOSTNAME=0.0.0.0
```

File: elements-cli.js [hunk 2]
```diff
@@ -21,6 +21,7 @@ const argd = {
   'base-path': process.env.ELEMENTS_BASE_PATH || process.env.BASE_PATH  || '/',
   hostname: process.env.ELEMENTS_HOSTNAME || 'localhost',
   layout: process.env.ELEMENTS_LAYOUT || process.env.LAYOUT || 'sidebar',
+  'listen-address': process.env.ELEMENTS_LISTEN_ADDRESS || '127.0.0.1',
```
Comment (schroedan) [inline elements-cli.js:24]:
```suggestion
  'virtual-host': process.env.ELEMENTS_VIRTUAL_HOST || 'localhost',
  'virtual-port': process.env.ELEMENTS_VIRTUAL_PORT || '8000',
```

File: elements-cli.js [hunk 3]
```diff
@@ -92,11 +93,12 @@ if (argv.help || argv._.length < 2 || !['export', 'preview'].includes(argv._[0])
           `  ${chalk.green('-c  --with-cors-proxy')}      Enable CORS proxy capabilities`,
           `  ${chalk.green('-f, --filter-internal')}      Filter out any content which has been marked as internal with x-internal`,
           `  ${chalk.green('-h, --help')}                 Display this help message`,
-          `  ${chalk.green('    --hostname=HOSTNAME')}    Server hostname ${chalk.yellow('[default: "' + argd.hostname + '"]')}`,
+          `  ${chalk.green('    --hostname=HOSTNAME')}    Reported url hostname ${chalk.yellow('[default: "' + argd.hostname + '"]')}`,
```
Comment (schroedan) [inline elements-cli.js:96]:
```suggestion
          `  ${chalk.green('    --hostname=HOSTNAME')}    Server hostname ${chalk.yellow('[default: "' + argd.hostname + '"]')}`,
```

Hint: Please also fix indention of options

File: elements-cli.js [hunk 4]
```diff
@@ -338,10 +340,15 @@ app.get(
 
 // Listen for HTTP connections
 
-const server = app.listen(argv.port, argv.hostname, () =>
+const server = app.listen(argv.port, argv['listen-address'], () =>
```
Comment (schroedan) [inline elements-cli.js:343]:
```suggestion
const server = app.listen(argv.port, argv.hostname, () =>
```

File: elements-cli.js [hunk 5]
```diff
@@ -92,11 +93,12 @@ if (argv.help || argv._.length < 2 || !['export', 'preview'].includes(argv._[0])
...
           `  ${chalk.green('-f, --filter-internal')}      Filter out any content which has been marked as internal with x-internal`,
           `  ${chalk.green('-h, --help')}                 Display this help message`,
-          `  ${chalk.green('    --hostname=HOSTNAME')}    Server hostname ${chalk.yellow('[default: "' + argd.hostname + '"]')}`,
+          `  ${chalk.green('    --hostname=HOSTNAME')}    Reported url hostname ${chalk.yellow('[default: "' + argd.hostname + '"]')}`,
           `  ${chalk.green('    --layout=LAYOUT')}        Layout for Elements: sidebar, stacked ${chalk.yellow('[default: "' + argd.layout + '"]')}`,
           `  ${chalk.green('    --logo=LOGO')}            URL of an image that will show as a small square logo next to the title`,
           `  ${chalk.green('-n  --no-try-it')}            Hide the "Try It" panel (the interactive API console)`,
           `  ${chalk.green('-p, --poll')}                 Use polling instead of file system events`,
+          `  ${chalk.green('    --listen-address=ADDR')}  Server listen address ${chalk.yellow('[default: ' + argd['listen-address'] + ']')}`,
```
Comment (schroedan) [inline elements-cli.js:101]:
```suggestion
          `  ${chalk.green('    --virtual-host=VIRTUAL_HOST')}  Reported hostname ${chalk.yellow('[default: ' + argd['virtual-host'] + ']')}`,
          `  ${chalk.green('    --virtual-port=VIRTUAL_PORT')}  Reported port ${chalk.yellow('[default: ' + argd['virtual-port'] + ']')}`,
```

File: elements-cli.js [hunk 6]
```diff
@@ -338,10 +340,15 @@ app.get(
 
 // Listen for HTTP connections
 
-const server = app.listen(argv.port, argv.hostname, () =>
+const server = app.listen(argv.port, argv['listen-address'], () =>
   console.error(
-    `Elements server listening on http://${argv.hostname}:${argv.port}${baseHref}`
+    `Elements server listening on ${argv['listen-address']}:${argv.port}`
```
Comment (schroedan) [inline elements-cli.js:345]:
```suggestion
    `Elements server listening on ${argv.hostname}:${argv.port}`
```

File: elements-cli.js [hunk 7]
```diff
@@ -338,10 +340,15 @@ app.get(
...
   )
+  if (argv.hostname) {
+    console.error(
+      `Visit http://${argv.hostname}:${argv.port}${baseHref}`
+    )
+  }
```
Comment (schroedan) [inline elements-cli.js:351]:
```suggestion
  console.error(
    `Visit http://${argv['virtual-host']}:${argv['virtual-port']}${baseHref}`
  )
```

**Review by schroedan — COMMENTED** [review]
Okay, now I understand the problem. You are concerned with the output in the console. I just don't want to change the `ELEMENTS_HOSTNAME`. So I suggest we introduce two variables `ELEMENTS_VIRTUAL_HOST=localhost` and `ELEMENTS_VIRTUAL_PORT=8000` and use them only for the output in the console.

**Comment by colinmollenhour** [comment]
That works for me, thanks! I committed your suggestions.

**Review by schroedan — APPROVED** [review]

**Comment by colinmollenhour** [comment]
Thanks for merging and thanks again for sharing your very useful tool!

**Pull request merged**